pydantic/_internal/_generate_schema.py

```diff
@@ -210,6 +211,42 @@ def modify_model_json_schema(
     return json_schema
 
 
+JsonEncoders = Dict[Type[Any], JsonEncoder]
+
+
+def _add_custom_serialization_from_json_encoders(
+    json_encoders: JsonEncoders | None, tp: Any, schema: CoreSchema
+) -> CoreSchema:
+    """Iterate over the json_encoders and add the first matching encoder to the schema.
+
+    Args:
+        json_encoders (JsonEncoders | None): A dictionary of types and their encoder functions.
+        tp (Any): The type to check for a matching encoder.
+        schema (CoreSchema): The schema to add the encoder to.
+    """
+    if json_encoders is None:
+        return schema
+    if 'serialization' in schema:
+        return schema
+    # Check the class type and its superclasses for a matching encoder
+    # Decimal.__class__.__mro__ (and probably other cases) doesn't include Decimal itself
+    for base in (tp, *tp.__class__.__mro__[:-1]):
+        try:
+            encoder = json_encoders[base]
+        except KeyError:
+            continue
+
+        serialization = schema.get('serialization') or core_schema.plain_serializer_function_ser_schema(encoder)
+        if 'ref' in schema:
+            return core_schema.definition_reference_schema(schema_ref=schema['ref'], serialization=serialization)  # type: ignore
```

Comment:
This seems problematic to me — it's not currently hit by any of the tests (as determined by putting `if 'ref' in schema: assert False`, and it seems like in principle it could result in a definition getting removed from the schema. Does that make sense? I'm trying to look further into it now, maybe it happens to be the case that this can't cause issues...

Reply:
I haven't been able to reproduce it (though investigating did lead me to the other comment I made, about self-referencing encoders), but yeah this does seem to make the assumption that the original definition will be present elsewhere, I'm not sure whether that is a safe assumption or not, maybe you know based on the way definitions get stored in `GenerateSchema` or whatever.